### PR TITLE
Fix ordering bugs for firewall rules

### LINKED ISSUE #302: Site to Site firewall idempotency
After using this module, I noticed it only marks the config as "changed" if the total number of rules changes. So in other words, adding or removing from the total lines of firewall policy will trigger a change. However, modifying the contents of a rule, for example changing an IP address of an existing dest, will not trigger a change. 

It looks like on line 284 the only check of the rules is based on length of the json.

Is there any plans to improve this check so that it validates the contents of the ruleset and not just the overall number of lines?

I can try to work on this but wasn't sure if there was already an updated file in the works. My python-fu may be a little rusty though.

https://github.com/CiscoDevNet/ansible-meraki/blob/b49a90872ffd690e34eebf2fb3eb933afd931daf/plugins/modules/meraki_mx_site_to_site_firewall.py#L284

### PR DESCRIPTION
**Pull request by kbreit**
Fixes #302

### CODE REVIEW

**Comment by michael-andretta** [comment]
How would I test against this branch specifically while I'm using the collection?

**Comment by kbreit** [comment]
I really should add this to the documentation. I didn't test it but you should be able to use something like this:

```
ansible-galaxy collection install git+https://github.com/CiscoDevNet/ansible-meraki.git,bugfix/302-firewall-rules
```

You may need to quote the path to escape the `/`. But that will install the branch collection. You may want it to go into a separate path so you can delete it later.

**Comment by michael-andretta** [comment]
> I really should add this to the documentation. I didn't test it but you should be able to use something like this:
> 
> ```
> ansible-galaxy collection install git+https://github.com/CiscoDevNet/ansible-meraki.git,bugfix/302-firewall-rules
> ```
> 
> You may need to quote the path to escape the `/`. But that will install the branch collection. You may want it to go into a separate path so you can delete it later.

TL;DR: This fix looks good. 

OK, so I wasn't able to get it to work this way. But I was able to test it. First off THANK YOU! Secondly, this is the error I was getting. 
```
ERROR! Invalid collection name 'git+https', name must be in the format <namespace>.<collection>. Please make sure namespace and collection name contains characters from [a-zA-Z0-9_] only.
```

What I ended up doing was modifying my role which called the FQCN of the collection/module, to call a local python file. I put that into my repo in library/meraki_mx_site_to_site_firewall.py inside were the contents of your fixed file here for testing. 

```
- name: Set Global Site-to-Site Firewall Rules
  # cisco.meraki.meraki_mx_site_to_site_firewall:
  meraki_mx_site_to_site_firewall:
    auth_key: '{{ api_key }}'
    org_name: '{{ full_org_name }}'
    state: present
    rules: '{{ meraki_global_firewal_policy }}'
    syslog_default_rule: yes
  delegate_to: localhost
```

Here is my vars file:
```
meraki_global_firewal_policy:
  - comment: First Rule
    src_cidr: 192.0.2.0/24
    src_port: any
    dest_cidr: 192.0.2.2/32, 192.0.3.0/24
    dest_port: '23, 22'
    protocol: tcp
    policy: deny
    syslog_enabled: true

  - comment: Second Rule
    src_cidr: 3.3.3.3/32
    src_port: any
    dest_cidr: 4.4.4.4/32
    dest_port: '9000'
    protocol: tcp
    policy: allow
    syslog_enabled: true

  - comment: Third Rule
    src_cidr: 192.0.1.0/24
    src_port: any
    dest_cidr: 192.0.2.2/32
    dest_port: '80, 443, 8443'
    protocol: tcp
    policy: allow
    syslog_enabled: true
```

Original Output:
image

Modified vars file:
```
meraki_global_firewal_policy:
  - comment: First Rule
    src_cidr: 192.0.2.0/24
    src_port: any
    dest_cidr: 192.0.2.2/32, 192.0.3.0/24
    dest_port: '23, 22'
    protocol: tcp
    policy: deny
    syslog_enabled: true

  - comment: Third Rule
    src_cidr: 192.0.1.0/24
    src_port: any
    dest_cidr: 192.0.2.2/32
    dest_port: '80, 443, 8443'
    protocol: tcp
    policy: allow
    syslog_enabled: true

  - comment: Second Rule
    src_cidr: 3.3.3.3/32
    src_port: any
    dest_cidr: 4.4.4.4/32
    dest_port: '9000'
    protocol: tcp
    policy: allow
    syslog_enabled: true
```

Final Results:

image

**Comment by michael-andretta** [comment]
One thing we did notice today while working with this, even with rules unchanged, the policy always shows changed. Not a deal breaker, but wanted to let you know.

**Comment by kbreit** [comment]
Thanks. I'll take a look at why it's showing changed. That shouldn't be.

**Comment by kbreit** [comment]
@michael-andretta If you would,  please try to Capitalize the A in `any`. I think it's a case sensitivity problem.

**Comment by kbreit** [comment]
I haven't tested enough to know exactly where it happens. But the example shows lowercase and the API appears to respond with `Any`. The comparison is detecting a mismatch there which is why it's showing there is always a change.

**Comment by michael-andretta** [comment]
> I haven't tested enough to know exactly where it happens. But the example shows lowercase and the API appears to respond with `Any`. The comparison is detecting a mismatch there which is why it's showing there is always a change.

Ill do some testing with various parameters that could possibly contain 'Any' with the variations and let you know.

**Comment by kbreit** [comment]
For now I am going to leave the `protocol` one as lowercase since that has options defined. I think it's odd to have `any` and `Any` in the documentation, purely for case sensitivity. If you want, I can open another issue and PR to investigate if it'll support any case for the `protocol` and see if I can make the inputs case insensitive.

The latest update to this branch normalizes the `any` for `src_*` and `dest_*`. Would you test it? Thank you.

**Comment by michael-andretta** [comment]
> For now I am going to leave the `protocol` one as lowercase since that has options defined. I think it's odd to have `any` and `Any` in the documentation, purely for case sensitivity. If you want, I can open another issue and PR to investigate if it'll support any case for the `protocol` and see if I can make the inputs case insensitive.
> 
> The latest update to this branch normalizes the `any` for `src_*` and `dest_*`. Would you test it? Thank you.

Just tested 2 iterations of this, they both show unchanged, looks good. 
```
  - comment: CLEANUP RULE
    src_cidr: any
    src_port: any
    dest_cidr: Any
    dest_port: Any
    protocol: any
    policy: deny
    syslog_enabled: true
```
and
```
  - comment: CLEANUP RULE
    src_cidr: any
    src_port: any
    dest_cidr: any
    dest_port: any
    protocol: any
    policy: deny
    syslog_enabled: true
```

**Comment by kbreit** [comment]
Thank you everyone for your assistance. I'm merging this.

**Comment by michael-andretta** [comment]
> Thank you everyone for your assistance. I'm merging this.

Thanks @kbreit, im still using the local script to get the desired results. When will this be published into a new tag like 2.15.2, or 2.16.1?

**Comment by kbreit** [comment]
I'll hopefully release a new version this week. It shouldn't take a lot of effort to release it.
